### PR TITLE
FLOPs (floating-point operations) module for ConvNets

### PR DESCRIPTION
**Pull request by ohaijen**
This code updates an existing but unused model analyzer (AnalyzerModule) object that computes forward FLOPs, parameters, prunable parameters, and zeroed parameters model-wide. Note that this is somewhat redundant with the [ModelPruningAnalyzer](https://github.com/neuralmagic/sparseml/blob/a0298c8b5d7dbcfd195d36a767d0b8bce1c45f25/src/sparseml/pytorch/optim/analyzer_pruning.py#L32), which is invoked by pruning modifiers. However, there are some key differences:

- The AnalyzerModule counts FLOPs as well as sparsities.
- The AnalyzerModule, if enabled,  also counts cumulative FLOPs, and works at the granularity of the model + its submodules. This means that it must be enabled from the very first step until the last. This is unlike ModelPruningAnalyzer, which is at the granularity of a specific pruner, and is active only when that pruner is active, and only on the parameters on which the pruner operates.
- For the moment, the AnalyzerModule only supports module types that appear in ConvNets, and should not be used with transformers.

However, as a future TODO, the AnalyzerModule can easily subsume the operations of the ModelPruningAnalyzer, and probably it's better-positioned in terms of being training-wide rather than pruner-specific, so it might be worth it to remove the latter in some future iteration.

Since the AnalyzerModule was not used before this change, it was updated to better fit the needs of the benchmarking project. As such, it no longer tracks things like param shapes, but only those metrics that are useful, like forward FLOPs and sparsities. The "total FLOPs" field is now used to track cumulative FLOPs, which were not tracked previously. The analyzer does not track backward FLOPs, because the standard implementation is to approximate Backward FLOPs ~= 2*Forward FLOPs; this calculation is not part of the code but would have to be performed manually, possibly with adjustments for pruners like Top-KAST, where the backward sparsity doesn't match the forward. The code is completely agnostic to this sort of nuance, which seems easier.

Note that the definitions of FLOP calculations have been changed to match those of AC/DC and RigL. In the vast majority of cases this did not affect the calculation results (which can be seen in the tests).

Testing: updated/expanded the existing tests + ran a few epochs to compare agreement with numbers published for AC/DC and RigL.

### CODE REVIEW

File: src/sparseml/pytorch/optim/analyzer_module.py [hunk 1]
```diff
@@ -333,28 +341,14 @@ def _bn_hook(
...
-            key: tuple(s for s in val.shape) for key, val in params.items()
-        }
-        desc.prunable_params_dims = {
-            key: tuple(s for s in val.shape) for key, val in prunable_params.items()
-        }
-
-        # 4 elementwise operations on the output space, just need to add all of them up
-        desc.flops = 4 * float(numpy.prod(out[0].shape[1:]))
-        desc.total_flops = desc.flops
+        desc.flops = 2 * float(inp[0].nelement())
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/optim/analyzer_module.py:350]: I'm confused as to why. The batchnorm takes every input element and normalizes it, no? This is also consistent with the implementations in AC/DC (https://github.com/IST-DASLab/ACDC/blob/ac53210b6adc1f2506ff909de08172ed9cad25d5/utils/flop_utils.py#L112) and the one AC/DC is based on (https://github.com/simochen/model-tools/blob/master/pytorch_tools.py#L61).

[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/optim/analyzer_module.py:350]: I left it out because we (the community) don't typically sparsify BN layers. But of course I can add it here, too.

File: src/sparseml/pytorch/optim/analyzer_module.py [hunk 2]
```diff
@@ -365,26 +359,20 @@ def _pool_hook(
...
-        desc.stride = mod.stride
+        desc.prunable_params = 0
+        desc.zeroed_params = 0
+
+        batch_size, input_channels, input_height, input_width = inp[0].size()
+        batch_size, output_channels, output_height, output_width = out[0].size()
 
-        flops_per_out_pix = float(numpy.prod(mod.kernel_size) + 1)
-        out_pix = float(numpy.prod(out[0].shape[1:]))
+        kernel_ops = mod.kernel_size * mod.kernel_size
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/optim/analyzer_module.py]: Fixed, thanks!

File: src/sparseml/pytorch/optim/analyzer_module.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmarques) [reply src/sparseml/pytorch/optim/analyzer_module.py:351]: Makes sense

[the remark this answers is not in the document]
Comment (anmarques) [reply src/sparseml/pytorch/optim/analyzer_module.py:351]: Yep, sounds good

**Review by anmarques — APPROVED** [review]

**Review by anmarques — APPROVED** [review]

**Review by shubhra — APPROVED** [review]
This looks good overall - if time permits, it would be great to add more comments to help make the calculations easier to follow through the logic.
